Add relation between import configurations and clients

solth opened the pull request:
Since import configurations can contain sensitive data like FTP credentials, they shouldn't be freely shared between clients. This pull request adds a relation between clients and import configurations similar to that between clients and projects, so that import configurations are explicitely mapped to specific lists of clients and users can only see and use those import configurations that are mapped to at least one of their assigned clients.

(since XSLT mapping files normally do _not_ contain such sensitive data and are - like rulesets - available to all users via the file system anyway, a relation between clients and mapping files does not seem necessary and therefore has not been added)

The mapping between import configurations and clients can be edited via a new third tab named "Mandanten" ("Clients") on the import configuration edit page:

<img width="437" alt="Bildschirmfoto 2024-07-12 um 14 05 53" src="https://github.com/user-attachments/assets/4c87051e-3450-4d58-b1f5-8dcdf5729397">

The SQL migration file in this pull request that adds the corresponding cross table also assigns _all_ existing import configurations to _all_ existing clients by default to maintain the current status quo. An admin can then make use of the new functionality instroduced in this pull request and restrict access to specific configurations using the import configuration edit formular shown above.

Fixes #6059

henning-gerhardt reviewed (commented):
Code changes look good, but I have 2 functional questions:

1. It is possible to de-assign all clients from an import configuration. This import configuration becomes invisible but is still existing. Without manipulating the database table it is not possible anymore to interact with this hidden import configuration. In my opinion at least one client should be assigned to an import configuration.

2. If you have at least two clients Client_A and Client_B. You logged in in Client_A and de-assign an import configuration from Client_A so this import configuration should only assigned to Client_B, the import configuration is still visible while logged in in Client_A (even after re-log in). This behaviour is different to all the other client-depending relations like projects, process templates, workflows, ... Should this be the case?

solth commented:
@henning-gerhardt thanks a lot for the remarks!

> 1. It is possible to de-assign all clients from an import configuration. This import configuration becomes invisible but is still existing. Without manipulating the database table it is not possible anymore to interact with this hidden import configuration. In my opinion at least one client should be assigned to an import configuration.

Thanks for pointing this out. Indeed, it would become difficult to re-assign an import configuration that is not assigned to any client anymore. I see two ways to approach this:
- either add a specdial permission/authority that allows users to see _all_ import configurations, no matter to which client(s) they are assigned. This way an "System admin" could be defined that 
- or add some frontend logic that disables and item in the client list when it is the last selected item, to prevent the situation you described, where a import configuration isn't assigned to any client anymore. This approach would have to take into account that a user might only be assigned to _some_ clients (not all of them) and an import configuration might still be assigned to a client to which the user is _not_ assigned

What would be the better solution? Or do you have an alternative idea how to resolve this issue?

> 2. If you have at least two clients Client_A and Client_B. You logged in in Client_A and de-assign an import configuration from Client_A so this import configuration should only assigned to Client_B, the import configuration is still visible while logged in in Client_A (even after re-log in). This behaviour is different to all the other client-depending relations like projects, process templates, workflows, ... Should this be the case?

This indeed sounds like a bug, e.g. wasn't intended. I will try to find the cause for this issue and resolve it.

henning-gerhardt commented:
I would go for the first option (special authority to "see" all available clients) even this behaviour is different then an other places like projects, process templates, ... but I think that sharing import configurations is more common than sharing projects, process templates, ... between clients.

henning-gerhardt commented:
After a discussion between @solth  and @henning-gerhardt we come to the following solution:
- adding a new permission/authority to see all clients inside the import configuration independent of which your current client is
- adding a custon validator which ensures that at least one client is assigned to an import configuration

solth commented:
@henning-gerhardt I updated the pull request as discussed: a new special permission was added that allows assigning import configurations to clients globally. The corresponding and new third "Clients" tab in the "Import configuration edit" page is now only rendered when the user has this new permission. On this third tab _all_ clients are displayed and offered for assignement (not just those assigned to the current user/admin). Additionally, I added a new validator that ensures at least one client is assigned when saving the import configuration.

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/constants/StringConstants.java:
```diff
@@ -15,5 +15,6 @@ public class StringConstants {
 
     public static final String COMMA_DELIMITER = ", ";
     public static final String SEMICOLON_DELIMITER = "; ";
+    public static final String SAVE = "editForm:save";
```
Is `SAVE` not to meaningless and may be used in wrong context? Maybe `EDITFORM_SAVE` can prevent this?

solth replied:
Yes, you are right. The name of the constant should match the content. When it's called just "SAVE", I, too, would expect it to contain a simple String "Save". I will change that to your suggestion.

henning-gerhardt reviewed (approved):
Changes looks good and new feature is working as defined.

henning-gerhardt reviewed (approved):
Thank you for adding a test for the new feature.
